oauthplugin/oidcprovider.go

```diff
@@ -0,0 +1,120 @@
+package oauthplugin
+
+import (
+	"context"
+	"errors"
+
+	oidc "github.com/coreos/go-oidc"
+	"golang.org/x/oauth2"
+)
+
+type oidcAuth struct {
+	issuer       string
+	oidcProvider *oidc.Provider
+	oauthConfig  oauth2.Config
+	oidcConfig   oidc.Config
+	verifier     *oidc.IDTokenVerifier
+}
+
+func newOidcIdProvider(ctx context.Context, backendConf *pluginConfig) (*oidcAuth, error) {
+	// The context passed to NewProvider must remain valid for as long as we
+	// want to be able to use this provider. See
+	// https://github.com/coreos/go-oidc/pull/176
+	oidcProvider, err := oidc.NewProvider(context.Background(), backendConf.Issuer)
+	if err != nil {
+		return nil, err
+	}
+
+	provider := &oidcAuth{
+		issuer:       backendConf.Issuer,
+		oidcProvider: oidcProvider,
+		oidcConfig: oidc.Config{
+			ClientID: backendConf.ClientID,
+		},
+	}
+
+	provider.oauthConfig = oauth2.Config{
+		ClientID:     backendConf.ClientID,
+		ClientSecret: backendConf.ClientSecret,
+		RedirectURL:  "urn:ietf:wg:oauth:2.0:oob",
+		Endpoint:     oidcProvider.Endpoint(),
+		Scopes:       []string{oidc.ScopeOpenID, "email"},
+	}
+
+	provider.verifier = oidcProvider.Verifier(&provider.oidcConfig)
+
+	return provider, nil
+}
+
+func (idp *oidcAuth) ValidateCode(ctx context.Context, code string, redirectURL string) (map[string]interface{}, error) {
+	// The OAuth2 spec requires that when exchanging the authorization code for
+	// tokens, the redirect_uri is provided exactly as it was in the
+	// authorization request that generated the authorization code.
+	//
+	// https://www.oauth.com/oauth2-servers/access-tokens/authorization-code-request/
+	// https://tools.ietf.org/html/rfc6749#section-4.1.3
+	//
+	// vault-login-oauth uses a different redirect URI each time, firstly
+	// because it listens on a random port, and secondly in accordance with
+	// best practice for native apps:
```

Comment:
I'm also used to the web app flow, and the RFC isn't terribly clear either. Aside from the port, is this because each flow is a dynamic instance of an Authorization server and per the RFC each instance of an Authorization server MUST have a unique redirect URI?

Reply:
I believe these go into the issues with more detail:

- http://openid.net/2016/07/16/preventing-mix-up-attacks-with-openid-connect/
- https://tools.ietf.org/html/draft-ietf-oauth-mix-up-mitigation-01

I admit I haven't completely absorbed these documents either.